api/tests.py

```diff
@@ -0,0 +1,193 @@
+        self.assertEqual(puzzle.url, "https://newurl.test/")
+
+        response = self.edit_puzzle(puzzle.pk, {"is_meta": True})
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.data["is_meta"], True)
+        self.assertEqual(len(response.data["tags"]), 1)
+        puzzle.refresh_from_db()
+        self.assertEqual(puzzle.is_meta, True)
+
+    def test_list_puzzles(self):
+        self.create_puzzle({"name": "test name", "url": TEST_URL})
+        self.create_puzzle({"name": "second test", "url": "https://secondtest.test/"})
+        self.assertEqual(Puzzle.objects.count(), 2)
+        response = self.list_puzzles()
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response.data), 2)
+        self.assertEqual(
+            {p["name"] for p in response.data}, {"test name", "second test"}
+        )
+
+    def test_add_answer(self):
+        self.create_puzzle({"name": "test name", "url": TEST_URL})
+        puzzle = Puzzle.objects.get()
+
+        response = self.create_answer(puzzle.pk, {"text": "ans"})
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        # This assumes no answer queue
+        puzzle.refresh_from_db()
+        self.assertEqual(puzzle.correct_answers(), ["ANS"])
+        self.assertEqual(puzzle.status, Puzzle.SOLVED)
+
+        response = self.create_answer(puzzle.pk, {"text": "ans"})
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        response = self.create_answer(puzzle.pk, {"text": ""})
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        response = self.create_answer(puzzle.pk, {"text": "answer two"})
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        puzzle.refresh_from_db()
+        self.assertEqual(puzzle.correct_answers(), ["ANS", "ANSWERTWO"])
+
+    def test_delete_answer(self):
+        self.create_puzzle({"name": "test name", "url": TEST_URL})
+        puzzle = Puzzle.objects.get()
+
+        self.create_answer(puzzle.pk, {"text": "ANSWER"})
+        self.create_answer(puzzle.pk, {"text": "ANSWER TWO"})
+
+        puzzle.refresh_from_db()
+        self.assertEqual(puzzle.status, Puzzle.SOLVED)
+        self.assertEqual(len(puzzle.correct_answers()), 2)
+        guesses = list(puzzle.guesses.all())
+
+        response = self.delete_answer(puzzle.pk, guesses[0].pk)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        puzzle.refresh_from_db()
+        self.assertEqual(puzzle.status, Puzzle.SOLVED)
+        self.assertEqual(len(puzzle.correct_answers()), 1)
+
+        response = self.delete_answer(puzzle.pk, guesses[1].pk)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        puzzle.refresh_from_db()
+        self.assertEqual(puzzle.status, Puzzle.SOLVING)
+        self.assertEqual(len(puzzle.correct_answers()), 0)
+
+    def test_edit_answer(self):
```

Comment:
Edge cases:
- Editing an answer to another answer that already exists
- Special characters / emojis

Reply:
We should definitely test these but I think the answers model tests do/should take care of these ones

Reply:
(though from the API side I think ideally we should still test that the API returns reasonable responses in these cases)